builtins: add missing UnwrapDatum call and change panics to error

asubiotto opened the pull request:
getNameForArg does not handle tree.DOidWrappers. The assumption in this code is
that the datum gets unwrapped before calling this code. This commit adds a
missing UnwrapDatum call and changes an unhandled type panic to an error
instead.

Release note (bug fix): a panic observed as "unexpected arg type
tree.DOidWrapper" has been fixed.

Closes #43166

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49601)
<!-- Reviewable:end -->

asubiotto commented:
Not sure how to add a test for this, would appreciate any suggestions.

jordanlewis commented:
DOidWrapper is only used for Names. Try passing a Name into one of these casts, maybe?

asubiotto commented:
Done

asubiotto commented:
Friendly ping

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49601#-:-M8l4yEfDDs9vjINTNEr:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/logictest/testdata/logic_test/pg_builtins, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49601#-M8l4qNE3hCpcq37FLRp:-M8l4qNE3hCpcq37FLRq:b-ov1hwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6e8aedfa1ce32cc22f0c75c5e82ffd1f6281721/pkg/sql/logictest/testdata/logic_test/pg_builtins#L18)):*
> ```
> 0
> 
> # Regression test for https://github.com/cockroachdb/cockroach/issues/49072.
> ```

nit: I think we tend to use short form `#49072`.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
LGTM, thanks for adding the test!

Commit:
builtins: add missing UnwrapDatum call and change panics to error

getNameForArg does not handle tree.DOidWrappers. The assumption in this code is
that the datum gets unwrapped before calling this code. This commit adds a
missing UnwrapDatum call and changes an unhandled type panic to an error
instead.

Release note (bug fix): a panic observed as "unexpected arg type
tree.DOidWrapper" has been fixed.

asubiotto reviewed (commented):
TFTRs

bors r=yuzefovich,jordanlewis

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49601#-:-M8oYZY1Er5BppCNa1n1:bor6uan)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->
